Implemented dynamic array and tests.

Robin-Bjornsdotter opened the pull request:
Features:
- vector_create - creates vector
- vector_append - adds given element to an end
- vector_pop - removes element from bottom
- vector_free
- vector_insert - inserts an element to a given position
- vector_remove - removes element from a given position
- vector_create_size - creates vector of a given size
- vector_copy - copies contents of a vector and creates a new one assigning copied values to it
- vector_print

Tests:
- test_vector_create
- test_vector_append
- test_vector_pop
- test_vector_create_size
- test_vector_insert
- test_vector_remove
- test_vector_copy

Commit:
Implemented dynamic array and tests.

Features:
* vector_create - creates vector
* vector_append - adds given element to an end
* vector_pop - removes element from bottom
* vector_free
* vector_insert - inserts an element to a given position
* vector_remove - removes element from a given position
* vector_create_size - creates vector of a given size
* vector_copy - copies contents of a vector and creates a new one assigning copied values to it
* vector_print

Tests:
* test_vector_create
* test_vector_append
* test_vector_pop
* test_vector_create_size
* test_vector_insert
* test_vector_remove
* test_vector_copy

disconnect3d commented:
Imho it would be better to rename `vector_copy` to `vector_clone`.

disconnect3d commented on vector/array.c:
```diff
@@ -0,0 +1,123 @@
+  arr->size = 0;
+  
+  return arr;
+}
```
Think of edge cases of `vector_create_size` ;).

disconnect3d commented on vector/array.c:
```diff
@@ -0,0 +1,123 @@
+  if (!arr)
+    return false;
+  
+  else if (arr->size == (arr->capacity-1)){
```
The parentheses around `arr->capacity-1` are redundant.

Read  [C reference's operator_precedence article](http://en.cppreference.com/w/c/language/operator_precedence).

However... you did it well. Whenever you don't know/remember the precedence, it is better to use parentheses ;).

disconnect3d commented on vector/array.c:
```diff
@@ -0,0 +1,123 @@
+	return false; 
+    }
+   
+    memmove((arr->array+position+1), (arr->array+position), (arr->size - position)*sizeof(data_t)); 
```
Redundant parentheses; The `,` is a delimiter for arguments, so there is no ambiguity here.

Also it is arguable whether parentheses make it more or less readable... It is up to you.

disconnect3d commented on vector/array.c:
```diff
@@ -0,0 +1,123 @@
+
+  array* vector_copy(array* arr){
+    
+    array* arr3 = vector_create_size(arr->capacity);
```
Think of `vector_create_size` edge cases.

disconnect3d commented on vector/array.c:
```diff
@@ -0,0 +1,123 @@
+  
+    arr3->capacity = arr->capacity;
+    arr3->size = arr->size;
+    memcpy(arr3->array, arr->array, (arr->size)*sizeof(data_t));
```
Redundant parentheses around `arr->size`. As stated [here](http://en.cppreference.com/w/c/language/operator_precedence) the:

```
++ --   Suffix/postfix increment and decrement  Left-to-right
()  Function call
[]  Array subscripting
.   Structure and union member access
->  Structure and union member access through pointer
(type){list}    Compound literal(C99)
```

operators have the highest (_not sure if that's the best word to describe it_) precedence.

Also `(arr)->(size*sizeof(data_t))` would be just ... super weird (how would the compiler know which field would you like to access?).

disconnect3d commented on vector/test.c:
```diff
@@ -0,0 +1,81 @@
+void test_vector_create_size(){
+  array* arr =  vector_create_size(36);
+  ASSERT(arr->capacity == 36);
+}
```
I'd also check for `arr->size == 0`.

disconnect3d replied:
Also lacks `free(arr)`.

disconnect3d replied:
Also, add another testcase for this function, that will check for `NULL` return.

How could we do that? Well... that's not so straightforward, but just pass a very big number, that would be bigger then someone's RAM + SWAP memory sizes. E.g. the maximum possible number for `size_t` type ([check here](http://stackoverflow.com/questions/3472311/what-is-a-portable-method-to-find-the-maximum-value-of-size-t)).

disconnect3d commented on vector/test.c:
```diff
@@ -0,0 +1,81 @@
+  ASSERT(arr->size == 0);
+
+  vector_free(arr);
+}
```
How can you know that this will pass - `ASSERT(arr != NULL);`?

Actually... you can't. Imagine a situation you are working on a big project, and there are tons of tests. Whenever you push to the remote git repository, the test build occurs.

Such "test build" launches the tests one after another. But... the machine that launches those tests, might also simultaneously launch other builds - e.g. tests for commits made by other people working in your team.

There might be a situation on which the machine would simply run OOM (out of memory) - e.g. if there is no limit for simultaneous builds. **In such situation, this test would fail on this assertion.**

And... it is hard to make this "work in all cases". You could for example **mock** (read [here](https://en.wikipedia.org/wiki/Mock_object) or [here](http://stackoverflow.com/questions/2665812/what-is-mocking)) the `malloc`, `free` and `realloc` (and eventually `calloc`) calls, so that they would work differently. 

This is doable e.g. thanks to `LD_PRELOAD` environment variable (read [here](http://stackoverflow.com/questions/426230/what-is-the-ld-preload-trick) or for real explanation - [here](http://man7.org/linux/man-pages/man8/ld.so.8.html) - under `LD_PRELOAD`).

I won't ask you to do that, because it is a total overkill for this task and I guess nobody does it (because another hard question would have to be answered - how to mock those calls properly, so the tests would always work?).

#### EDIT

Of course if out of memory happens on the machine, some or even all of the tests will fail and that is acceptable. If such situation happens, you can't really do anything else then just accept it.

However you could of course try to decrease the possibility of occurence of such situation - simply by adding a limit of simultaneous builds and have a queue for the builds ;).

disconnect3d commented on vector/array.c:
```diff
@@ -0,0 +1,123 @@
+  arr->array[index] = 0;
+  --arr->size; 
+  return ret_el;
+}
```
What if the array is empty? :(

Commit:
Following fixes and additions:

renamed vector_copy to vector_clone
removed unnecessary parentheses
fixed test_vector_create_size
fixed vector_pop for the case when array is empty
added test_vector_create_size_NULL_return
fixed vector_create, vector_create_size and vector_copy

disconnect3d commented on vector/test.c:
```diff
@@ -78,4 +78,8 @@ void test_vector_copy(){
 
-
+void test_vector_create_size_NULL_return(){
+  size_t max_size = (size_t)-1;
```
`(size_t)-1` should return maximum value of `size_t` as it is an unsigned type...
However isn't the `SIZE_MAX` macro working? It is **much more readable**.

disconnect3d commented on vector/array.c:
```diff
@@ -0,0 +1,138 @@
+  
+      return arr;
+    }
+  }
```
There are some wrong indents up from here (seems like you pressed TAB on part of the code).

Commit:
added one assertion in test_vector_create_size

disconnect3d commented on vector/array.h:
```diff
@@ -0,0 +1,26 @@
+#ifndef array_h
+#define array_h
```
[Include guards](https://en.wikipedia.org/wiki/Include_guard), nice.

However, they have a possibility of **name clashing**. If you happen to have two headers with the same include guard, you will end up with weird compilation results which may take you some time till you understand what is going on (well `-E` gcc flag would show what is going on - or rather what have been included and what not, but it might not come up to one's mind).

...this is also described in the above linked wiki page as [**difficulties**](https://en.wikipedia.org/wiki/Include_guard#Difficulties).

Also, there is a [convention to write macros UPPERCASE](http://stackoverflow.com/questions/3799478/c-ifndef-for-include-files-why-is-all-caps-used-for-the-header-file), so such ifndef/define's are usually written uppercase as well.

There is another solution to the problem of including a file multiple times, which is **pragma once**. This is **NON-STANDARD** solution, but most of the compilers actually support it.

Read more here:
- https://en.wikipedia.org/wiki/Pragma_once
- http://stackoverflow.com/questions/1143936/pragma-once-vs-include-guards

Commit:
Changed indents and include guards names to uppercase

Commit:
Added makefile

Commit:
Added CLion project directory to .gitignore

Commit:
Refactored code

Commit:
added CMakeLists

disconnect3d commented:
The `position` argument used in the functions can be negative which will result in Undefined Behavior (e.g. `memset` to the address before `arr->array`).

So... maybe change its type to `size_t` (so that you won't have to check if it is negative).

disconnect3d commented on vector/array.c:
```diff
@@ -0,0 +1,132 @@
+    return true;
+}
+
+bool vector_remove(array *arr, int position) {
```
Add a testcase:

``` C
void test_vector_remove_empty_array() {
    array* arr = array_create();

    ASSERT(vector_remove(arr, 0) == false);
    ASSERT(arr->size == 0);
}
```

disconnect3d commented on vector/array.c:
```diff
@@ -0,0 +1,132 @@
+        return -1;
+    size_t index = arr->size - 1;
+    data_t ret_el = arr->array[index];
+    arr->array[index] = 0;
```
This will result in `arr->array[-1]` (which is UB) if the array is empty...

Commit:
added testcase test_vector_remove_empty_array

changed `int position` to `size_t position`
added `if` for handling a case in vector_pop when the array is empty

Commit:
removed `#include "array.c"

Commit:
Changed include guards to pragma once

Commit:
Changed vector_pop

now returns true/false whether popping an element was succesful and
passes popped value using a pointer.

disconnect3d commented on vector/array.c:
```diff
@@ -0,0 +1,134 @@
+        return false;
+    size_t index = arr->size - 1;
+    *popped_el = arr->array[index];
+    arr->array[index] = 0;
```
This is redundant.

(There are use cases for it, like when you would wipe out secret password from the memory or sth like that, but we don't need that in the `vector` library itself).

disconnect3d commented on vector/test.c:
```diff
@@ -0,0 +1,93 @@
+    data_t popped_el = 0;
+    vector_pop(arr, &popped_el);
+    ASSERT(popped_el == 3);
+
```
I'd add `ASSERT(arr->size == 2)` here.

disconnect3d commented on vector/test_main.c:
```diff
@@ -0,0 +1,13 @@
+    test_vector_remove();
+    test_vector_copy();
+    test_vector_create_size_NULL_return();
+    return 0;
```
One test call is missing.

Commit:
Added assertion in test_pop and removed redundant line in vector_pop

Commit:
Added missing test call

disconnect3d commented on vector/test.c:
```diff
@@ -0,0 +1,94 @@
+}
+
+void test_vector_create_size_NULL_return() {
+    size_t max_size = (size_t) -1   ;
```
Hmm... I was wrong. This can't be `SIZE_MAX` here. It must be `SIZE_MAX` divided by the size of data.

Because otherwise, valgrind says this:

```
==4874== Argument 'size' of function malloc has a fishy (possibly negative) value: -4
==4874==    at 0x4C2BBCF: malloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==4874==    by 0x400A5F: vector_create_size (array.c:75)
==4874==    by 0x401340: test_vector_create_size_NULL_return (test.c:85)
==4874==    by 0x4007C9: main (test_main.c:11)
==4874== 
```

Commit:
Fixed vector_remove case when there are no elements

disconnect3d commented on vector/array.c:
```diff
@@ -108,7 +108,8 @@ bool vector_insert(array *arr, data_t element, size_t position) {
         return false;
-
+    if (arr->size == 0)
+        return false;
```
Or just:

``` C
if (!arr || arr->size == 0)
    return false;
```

Commit:
Fixed test_vector_create_size_NULL_return
Changed makefile
Small changes in vector_remove
Added missing vector_frees in test.c
